### DISCUSSION TITLE
Allow DirectionalLight3D to respond realistically to atmospheric scattering

### DISCUSSION THREAD

**Discussion by Ophiolith** [godotengine/godot-proposals#3984] [3D]
This is a continuation of #3981 , which I closed as I realised it didn't meet the criteria for a full proposal.

Due to the way Godot uses the DirectionalLight3D node to control the rotation, colour and energy of the sun for sky materials, PhysicalSky and ProceduralSky, this produces very discontinuous lighting when the sun is angled close to the horizon, as can be seen below:

![image](https://user-images.githubusercontent.com/24281932/154315319-22e14646-d8ed-4ae6-9b86-3e80595eb8b1.png)

Because atmospheric scattering is handled entirely in the sky shader, there is no feasible way to pass back any modulation to the controlling light. I lack the technical expertise to implement a fix, so I'm opening this discussion to propose various means to potentially solve this. My thoughts are the following could be feasible:

- Implement a new light type called 'PhysicalSunLight3D' or similar, which precalculates the colouring caused by atmospheric scattering based on angle of rotation, and then have that information passed to the shader (basically, calculate direct light scattering on the light, not the shader).
- Implement a toggle that would prevent a light passing colour and energy to the shader, and allow users to set discrete lights for controlling the sun and the actual directional light (originally suggested by @clayjohn in the closed proposal).
- Have PhysicalSky simply not calculate atmospheric scattering colouration and leave it up to the user to script this via changes to DirectionalLight3D's properties.

Apologies for the lack of technical nous on this one, but it's been bugging me a lot recently. I'd appreciate any and all input on how this is best achieved.

**Comment by clayjohn** [comment]
> Implement a toggle that would prevent a light passing colour and energy to the shader, and allow users to set discrete lights for controlling the sun and the actual directional light.

I commented on the original proposal to suggest essentially this. It was also proposed in response to a bug report on the main Godot repo https://github.com/godotengine/godot/issues/53985

**Reply by Ophiolith** [reply]
Thanks! I had tried to locate similar issues via search before posting but was unsuccessful. Apologies for not correctly crediting the suggestion on this thread, I've now amended the original post.

**Reply by clayjohn** [reply]
No apology necessary, I think you opened this while I was still typing it out!

**Comment by Ophiolith** [comment]
[I've opened a new issue](https://github.com/godotengine/godot/issues/58206) on the main Godot Github repo that would at least partially alleviate this issue if implemented.
